graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+      Hint hint = fieldHint.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the default max age in the
+      // next step.
+      // This does allow setting the cache scope though.
+      //
+      // Note: setting both maxAge and inheritMaxAge doesn't make sense.
+      if (hint.getInheritMaxAge().orElse(false) && !fieldPolicy.hasMaxAge()) {
+        inheritMaxAge = true;
+        fieldPolicy.replace(hint.getScope());
+      } else {
+        fieldPolicy.replace(hint);
+      }
+    }
+
+    // If this field returns a composite type or is a root field and
+    // we haven't seen an explicit maxAge hint, set the maxAge to 0
+    // (uncached) or the default if specified in the constructor.
+    // (Non-object fields by default are assumed to inherit their
+    // cacheability from their parents. But on the other hand, while
+    // root non-object fields can get explicit hints from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedReturnType instanceof GraphQLCompositeType && !inheritMaxAge)
+            || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.restrictCachePolicy(fieldPolicy);
+        super.onCompleted(result, t);
+      }
+    };
+  }
+
+  @Override
+  public CompletableFuture<ExecutionResult> instrumentExecutionResult(
+      ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+    final CacheControlState state = parameters.getInstrumentationState();
+
+    if (!state.getOverallPolicyString().isPresent()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    if (!executionResult.getErrors().isEmpty()) {
+      return super.instrumentExecutionResult(executionResult, parameters);
+    }
+
+    ExecutionResultImpl resultWithCacheControl =
+        ExecutionResultImpl.newExecutionResult()
+            .from(executionResult)
+            .addExtension(EXTENSION_KEY, state.getOverallPolicyString().get())
+            .build();
+
+    return CompletableFuture.completedFuture(resultWithCacheControl);
+  }
+
+  enum CacheControlScope {
+    PUBLIC,
+    PRIVATE
+  }
+
+  private static class CacheControlState implements InstrumentationState {
+    private final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+
+    public Optional<String> getOverallPolicyString() {
```

Comment:
If you think that including `String` in this method name is helpful, then perhaps it would also be helpful on the method that it calls. (That said, since this entire class is private, maybe just make `overallPolicy` be a public final field and skip the methods?)